apollo-router/src/plugins/telemetry/mod.rs

```diff
@@ -585,6 +482,136 @@ impl Plugin for Telemetry {
 }
 
 impl Telemetry {
+    /// This method can be used instead of `Plugin::new` to override the subscriber
+    pub async fn new_with_subscriber<S>(
+        init: PluginInit<<Self as Plugin>::Config>,
+        subscriber: Option<S>,
+    ) -> Result<Self, BoxError>
+    where
+        S: Subscriber + Send + Sync + for<'span> LookupSpan<'span>,
+    {
+        // Apollo config is special because we enable tracing if some env variables are present.
+        let mut config = init.config;
+        let apollo = config
+            .apollo
+            .as_mut()
+            .expect("telemetry apollo config must be present");
+
+        // If we have key and graph ref but no endpoint we start embedded spaceport
+        let (spaceport, shutdown_tx) = match apollo {
+            apollo::Config {
+                apollo_key: Some(_),
+                apollo_graph_ref: Some(_),
+                endpoint: None,
+                ..
+            } => {
+                ::tracing::debug!("starting Spaceport");
+                let (shutdown_tx, shutdown_rx) = futures::channel::oneshot::channel();
+                let report_spaceport = ReportSpaceport::new(
+                    "127.0.0.1:0".parse()?,
+                    Some(Box::pin(shutdown_rx.map(|_| ()))),
+                )
+                .await?;
+                // Now that the port is known update the config
+                apollo.endpoint = Some(Url::parse(&format!(
+                    "https://{}",
+                    report_spaceport.address()
+                ))?);
+                (Some(report_spaceport), Some(shutdown_tx))
+            }
+            _ => (None, None),
+        };
+
+        // Setup metrics
+        // The act of setting up metrics will overwrite a global meter. However it is essential that
+        // we use the aggregate meter provider that is created below. It enables us to support
+        // sending metrics to multiple providers at once, of which hopefully Apollo Studio will
+        // eventually be one.
+        let mut builder = Self::create_metrics_exporters(&config)?;
+
+        // the global tracer and subscriber initialization step must be performed only once
```

Comment:
I suppose having multiple instances of this plugin (or multiple routers) in the same process is not supported?

Reply:
Yeah, it's a no go because of the globals within otel and tracing crates.

Reply:
Should we enforce this by using `once_cell.set(_).unwrap()` instead of `once_cell.get_or_try_init(|| _)`?

Reply:
no, there will be multiple instances of the plugin, because a new one is created when we reload the configuration, then the old one is removed. If we want to enforce only one instance, then telemetry should be outside of the plugin lifecycle and would not make sense as a plugin.

The issue with multiple instances is worked around with refcounting in 182c97d8 for provider shutdown